libtiledbsoma/src/utils/fastercsx.h

```diff
@@ -0,0 +1,545 @@
+ * @returns A std::pair containing the half-open range for the section.
+ */
+template <typename T>
+std::pair<T, T> get_split_(T n_elements, T n_sections, T section) noexcept {
+    auto base_size = n_elements / n_sections;
+    auto extras = n_elements % n_sections;
+
+    auto start = section * base_size + std::min(section, extras);
+    auto stop = (section + 1) * base_size + std::min(section + 1, extras);
+
+    return std::pair<T, T>(start, stop);
+}
+
+#ifndef NDEBUG
+/**
+ * @brief Sum size over all elements in vector - used only in asserts.
+ */
+template <typename T>
+size_t sum_over_size_(const std::vector<tcb::span<const T>>& v) noexcept {
+    return std::transform_reduce(
+        v.cbegin(), v.cend(), 0ul, std::plus<>{}, [](tcb::span<const T> a) {
+            return a.size();
+        });
+}
+#endif
+
+/**
+ * @brief Simple partitioning scheme - given a vector of spans, and a minimum
+ * partition size, create up to max_partitions.
+ *
+ * Used to split work across workers in a reasonably even manner. Currently a
+ * greedy n-way linear partition, which works well enough for the common case
+ * where most inputs are of similar sizes (which is typically the case).
+ *
+ * TODO: If we sort (descending) first, it will be classic LPT (Largest
+ * Processing Time) partitioning which would be better in edge cases where there
+ * are a lot of variable sized input vectors.
+ */
+template <typename COO_IDX>
+struct Partition {
+    size_t size;  // sum of views[n].size for all n
+    std::vector<tcb::span<COO_IDX>> views;
+};
+
+template <typename COO_IDX>
+void bin_view_(
+    std::vector<Partition<COO_IDX>>& partitions,
+    const tcb::span<COO_IDX>& view) {
+    // find minimum size partition and add the span to that partition.
+    size_t min_idx = 0;
+    for (size_t pidx = 1; pidx < partitions.size(); ++pidx) {
+        if (partitions[pidx].size < partitions[min_idx].size)
+            min_idx = pidx;
+    }
+    partitions[min_idx].size += view.size();
+    partitions[min_idx].views.push_back(view);
+}
+
+template <typename COO_IDX>
+std::vector<Partition<COO_IDX>> partition_views_(
+    std::vector<tcb::span<COO_IDX>> const& Ai,
+    const size_t max_partitions,
+    const size_t partition_size) {
+    assert(max_partitions > 0);
+    std::vector<Partition<COO_IDX>> partitions(max_partitions);
+    for (auto& view : Ai) {
+        size_t n_partitions = std::min(
+            (view.size() + partition_size - 1) / partition_size,
+            max_partitions);
+        for (size_t i = 0; i < n_partitions; ++i) {
+            const auto [start, stop] = get_split_(view.size(), n_partitions, i);
+            bin_view_(partitions, view.subspan(start, stop - start));
+        }
+    }
+
+    // Erase any partitions that are zero length
+    for (auto it = partitions.begin(); it != partitions.end();) {
+        if (it->size == 0)
+            it = partitions.erase(it);
+        else
+            ++it;
+    }
```

Comment:
```suggestion
    partitions.erase(
        std::remove_if(partitions.begin(), partitions.end(),
                       [](const Partition& p) { return p.size == 0; }),
        partitions.end());
```

Reply:
ah, good catch!  If only we could use `erase_if` :-)